toolkit/tools/imagegen/configuration/packagerepo.go

```diff
@@ -0,0 +1,217 @@
+
+// Parser for the image builder's PackageRepo configuration schemas.
+
+package configuration
+
+import (
+	"encoding/json"
+	"fmt"
+	"net/url"
+	"os"
+	"path/filepath"
+	"strings"
+
+	"github.com/microsoft/CBL-Mariner/toolkit/tools/internal/file"
+	"github.com/microsoft/CBL-Mariner/toolkit/tools/internal/logger"
+	"github.com/microsoft/CBL-Mariner/toolkit/tools/internal/network"
+	"github.com/microsoft/CBL-Mariner/toolkit/tools/internal/safechroot"
+)
+
+// PackageRepo defines the RPM repo to pull packages from during the installation
+// or after the system is installed. The "Install" option indicates that the provided
+// repository configuration will be saved in the installed system if specified, and only
+// available during the installation process if not
+type PackageRepo struct {
+	Name    string `json:"Name"`
+	BaseUrl string `json:"BaseUrl"`
+	Install bool   `json:"Install"`
+}
+
+// UnmarshalJSON Unmarshals a PackageRepo entry
+func (p *PackageRepo) UnmarshalJSON(b []byte) (err error) {
+	// Use an intermediate type which will use the default JSON unmarshal implementation
+	type IntermediateTypePackageRepo PackageRepo
+	err = json.Unmarshal(b, (*IntermediateTypePackageRepo)(p))
+	if err != nil {
+		return fmt.Errorf("failed to parse [PackageRepo]: %w", err)
+	}
+
+	// Now validate the resulting unmarshaled object
+	err = p.IsValid()
+	if err != nil {
+		return fmt.Errorf("failed to parse [PackageRepo]: %w", err)
+	}
+	return
+}
+
+// IsValid returns an error if the PackageRepo struct is not valid
+func (p *PackageRepo) IsValid() (err error) {
+	err = p.nameIsValid()
+	if err != nil {
+		return
+	}
+
+	err = p.repoUrlIsValid()
+	if err != nil {
+		return
+	}
+
+	return
+}
+
+// UpdatePackageRepo creates additional repo files specified by image configuration
+// and returns error if the operation fails
+func UpdatePackageRepo(installChroot *safechroot.Chroot, config SystemConfig) (err error) {
+	const (
+		repoFileDir   = "/etc/yum.repos.d/"
+		localRepoFile = "/etc/yum.repos.d/mariner-iso.repo"
+		squashErrors  = false
+	)
+
+	if exists, ferr := file.DirExists(repoFileDir); ferr != nil {
+		logger.Log.Errorf("Error accessing repo file directory %s.", repoFileDir)
+		err = ferr
+		return
+	} else if !exists {
+		err = fmt.Errorf("Could not find the repo file directory %s to update package repo", repoFileDir)
+		return
+	}
+
+	// Remove mariner-iso.repo
+	os.Remove(localRepoFile)
+
+	// Loop through the PackageRepos field to determine if any customized package repos are specified.
+	// If specified, create new repo files for them
+	for _, packageRepo := range config.PackageRepos {
+		err = createCustomPackageRepo(installChroot, packageRepo, repoFileDir)
+		if err != nil {
+			return
+		}
+	}
+
+	// It is possible that network access may not be up at this point,
+	// so check network access
+	err, hasNetworkAccess := network.CheckNetworkAccess()
+	if !hasNetworkAccess {
+		err = fmt.Errorf("no network access in the system")
+	}
```

Comment:
Awesome, thanks for the change. We can now distinguish between not having any network access and not being able to check if it's available:

```suggestion
	if err != nil {
		return
	}
	if !hasNetworkAccess {
		err = fmt.Errorf("no network access in the system")
	}
```